utils/python/CIME/utils.py

```diff
@@ -187,17 +177,29 @@ def run_cmd(cmd, ok_to_fail=False, input_str=None, from_dir=None, verbose=None,
         if errput:
             logger.info("  errput: %s\n" % errput)
 
-    if (ok_to_fail):
-        return stat, output, errput
-    else:
-        if (arg_stderr is not None):
-            errput = errput if errput is not None else open(arg_stderr.name, "r").read()
-            expect(stat == 0, "Command: '%s' failed with error '%s'%s" %
-                   (cmd, errput, "" if from_dir is None else " from dir '%s'" % from_dir))
-        else:
-            expect(stat == 0, "Command: '%s' failed%s. See terminal output" %
-                   (cmd, "" if from_dir is None else " from dir '%s'" % from_dir))
-        return output
+    return stat, output, errput
+
+def run_cmd_no_fail(cmd, input_str=None, from_dir=None, verbose=None,
```

Comment:
Important change. I split run_cmd into two functions, a fully generic run_cmd, and run_cmd_no_fail. Run_cmd_no_fail only returns the command output because it expects the command to succeed. An exception is raised if the command does not succeed. The previous implementation caused run_cmd to have different return types based on the ok_to_fail flag which is poor design and confused pylint.